libraries/tarball.rb

```diff
@@ -0,0 +1,157 @@
+      @options = options
+    end
+
+    attr_reader :prefix
+
+    def get_file(name)
+      data = []
+      target_path = @prefix.nil? ? name : "#{@prefix}/#{name}"
+      iterate file: proc { |path, entry| data << entry.read if path == target_path }, other: proc {}
+      fail "Multiple entries for #{target_path} found in #{@file_name}" if data.size > 1
+      data.first
+    end
+
+    def extract(extract_dir)
+      file_handler = proc do |path, entry|
+        target = "#{extract_dir}/#{path}"
+        make_dirs extract_dir, path
+        File.open(target, 'wb') { |f| f.print entry.read }
+        FileUtils.chmod entry.header.mode, target
+        FileUtils.chown @options[:user], @options[:group], target
+      end
+
+      dir_handler = proc do |path, entry|
+        target = "#{extract_dir}/#{path}"
+        make_dirs extract_dir, path, true
+        FileUtils.chmod entry.header.mode, target
+      end
+
+      symlink_handler = proc do |path, entry, other = {}|
+        target = "#{extract_dir}/#{path}"
+        make_dirs extract_dir, path
+        File.unlink target if File.exist? target
+        File.symlink other[:linkname], target
+        FileUtils.chmod entry.header.mode, target
+        FileUtils.chown @options[:user], @options[:group], target
+      end
+
+      iterate file: file_handler, directory: dir_handler, symlink: symlink_handler
+    end
+
+    # Count the number of (qualified) entries in the tarball where the return value
+    # of the corresponding callback applied to each entry in the tarfile is truthy.
+    # With no arguments, counts the number of qualified entries
+    # If called with a block (only), the block is the callback to for every entry.
+    # If called with a map, invoke the appropriate callback for each corresponding entry.
+    # See iterate for details on how callbacks are invoked.
+    def count(callbacks = {}, &other)
+      callbacks = { other: other } if other && callbacks.empty?
+      i = 0
+      # Wrap each callback such that if it's truthy, we increment our counter.
+      hsh = callbacks.each_with_object({}) { |(k, v), h| h[k] = proc { |*args| i += 1 if v.call(args) } }
+      hsh[:other] = proc { i += 1 } if hsh.empty?
+      hsh[:other] = proc {} unless hsh[:other]
+      iterate hsh
+      i
+    end
+
+    def num_files
+      count file: proc { true }
+    end
+
+    def make_dirs(extract_dir, path, all = false)
+      segments = path.split('/')
+      segments = segments[0...-1] unless all
+      segments.each_with_object(extract_dir.dup) do |segment, target|
+        target << "/#{segment}"
+        FileUtils.mkdir target unless File.directory? target
+        FileUtils.chown @options[:user], @options[:group], target
+      end
+    end
+
+    # Iterate over the entries of the tarfile, and invoke the corresponding callback with each entry.
+    # We handle the LongFileName and LongSymLink entries as found (no callbacks, but they change values sent with other callbacks)
```

Comment:
Which indirection(s) specifically are you questioning?

1. I put this library together, to eliminate declaring a dependency on a native tar package being present (and avoid building indirection to work natively on both Windows and *nix systems.)
1. LongFileName (L) and LongSymLink (K) are separate entries in GNU tar files that exist if the subsequent "file" (file here being in terms of *nix everything is a file, so could be a file, symlink, directory, or something else) path or symlink target are over 99 ASCII characters long (100 characters with null). POSIX header struct: http://www.gnu.org/software/tar/manual/html_node/Standard.html
1. The indirection of including only those with a certain prefix is so that we could extract a single app if the tarfile contains multiple apps. Also gives us an early catch of the tarfile doesn't contain the app you think it does.
1. The indirection of providing callbacks is similar to the Visitor / Iterator pattern, where I can declare each operation (extract, find file, count) simply in terms of what do I do when I encounter a particular type of entry (:file, :directory, :symlink) or any other entry (:other) instead of reimplementing the same iteration and prefix filter logic over and over again.